src/Chapter3.hs

```diff
@@ -1006,6 +1185,71 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- Use different names to avoid conflicts with older code
+data WarriorActions 
+	= Attack
+	| DrinkPotion
+	| CastSpell
+
+data BeastActions
+	= Bite
+	| RunAway
+
+data Warrior = Warrior
+	{ warriorAttack :: Int
+	, warriorActSeq :: [WarriorActions]
+	, warriorDefense :: Int
+	, warriorHealth :: Int
+	, warriorActIndex :: Int
+	}
+
+data Beast = Beast
+	{ beastAttack :: Int
+	, beastActSeq :: [BeastActions]
+	, beastHealth :: Int
+	, beastActIndex :: Int
+	}
+
+class Fighter a where
+	action :: a -> a
+	getDamage :: a -> Int -> a
+
+instance Warrior where
+	act :: Warrior -> Warrior
+	act a = case (warriorActSeq a) of
+		Attack = a
+		CastSpell = a { warriorDefense = (warriorDefense a) + 10 }
+		DrinkPotion = a { warriorHealth = (warriorHealth a) + 10 }
+	getDamage :: Warrior -> Int -> Warrior
+	getDamage a d = a { warriorHealth = (warriorHealth a) - getDelta a d }
+		where
+			getDelta :: Warrior -> Int -> Int
+			getDelta w dmg
+				| dmg < (warriorDefense w) = 0
+				| otherwise = dmg - (warriorDefense w)
+
+instance Beast where
+	act :: Beast -> Beast
+	act a = a
```

Comment:
It looks like the `act` action for `Beast` doesn't handle the `RunAway` action :thinking: 
Maybe `act` function should return `Maybe a` to handle this case? Just an idea :slightly_smiling_face: